File: jdk/src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLLayer.m
/*
 * Copyright 2018 JetBrains s.r.o.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

#import "MTLGraphicsConfig.h"
#import "MTLLayer.h"
#import "ThreadUtilities.h"
#import "LWCToolkit.h"
#import "MTLSurfaceData.h"


static const int N = 2;
static struct Vertex verts[N*3];

@implementation MTLLayer


@synthesize javaLayer;
@synthesize ctx;
@synthesize bufferWidth;
@synthesize bufferHeight;

- (id) initWithJavaLayer:(JNFWeakJObjectWrapper *)layer
{
//fprintf(stderr, "MTLayer.initWithJavaLayer\n");
AWT_ASSERT_APPKIT_THREAD;
    // Initialize ourselves
    self = [super init];
    if (self == nil) return self;

    self.javaLayer = layer;

    self.contentsGravity = kCAGravityTopLeft;

    //Disable CALayer's default animation
    NSMutableDictionary * actions = [[NSMutableDictionary alloc] initWithObjectsAndKeys:
                                    [NSNull null], @"anchorPoint",
                                    [NSNull null], @"bounds",
                                    [NSNull null], @"contents",
                                    [NSNull null], @"contentsScale",
                                    [NSNull null], @"onOrderIn",
                                    [NSNull null], @"onOrderOut",
                                    [NSNull null], @"position",
                                    [NSNull null], @"sublayers",
                                    nil];
    self.actions = actions;
    [actions release];

    self.ctx = NULL;

    return self;
}



- (void) blitTexture {
    if (self.ctx == NULL) {
        return;
    }
fprintf(stderr, "blitTexture\n");
    if (ctx->mtlCommandBuffer) {
    // Encode render command.
        id<CAMetalDrawable> mtlDrawable = [self nextDrawable];
        if (!ctx->mtlRenderPassDesc) {
            ctx->mtlRenderPassDesc = [[MTLRenderPassDescriptor renderPassDescriptor] retain];
        }

        if (ctx->mtlRenderPassDesc) {

            verts[0].position[0] = -1.0;
            verts[0].position[1] = 1.0;
            verts[0].position[2] = 0;
            verts[0].txtpos[0] = 0;
            verts[0].txtpos[1] = 0;

            verts[1].position[0] = 1.0;
            verts[1].position[1] = 1.0;
            verts[1].position[2] = 0;
            verts[1].txtpos[0] = 1;
            verts[1].txtpos[1] = 0;

            verts[2].position[0] = 1.0;
            verts[2].position[1] = -1.0;
            verts[2].position[2] = 0;
            verts[2].txtpos[0] = 1;
            verts[2].txtpos[1] = 1;


            verts[3].position[0] = 1.0;
            verts[3].position[1] = -1.0;
            verts[3].position[2] = 0;
            verts[3].txtpos[0] = 1;
            verts[3].txtpos[1] = 1;


            verts[4].position[0] = -1.0;
            verts[4].position[1] = -1.0;
            verts[4].position[2] = 0;
            verts[4].txtpos[0] = 0;
            verts[4].txtpos[1] = 1;

            verts[5].position[0] = -1.0;
            verts[5].position[1] = 1.0;
            verts[5].position[2] = 0;
            verts[5].txtpos[0] = 0;
            verts[5].txtpos[1] = 0;

            ctx->mtlVertexBuffer = [ctx->mtlDevice newBufferWithBytes:verts
                                                           length:sizeof(verts)
                                                           options:MTLResourceCPUCacheModeDefaultCache];
            MTLRenderPassColorAttachmentDescriptor *colorAttachment = ctx->mtlRenderPassDesc.colorAttachments[0];

            colorAttachment.texture = mtlDrawable.texture;

            colorAttachment.loadAction = MTLLoadActionLoad;
            colorAttachment.clearColor = MTLClearColorMake(0.0f, 0.0f, 0.0f, 1.0f);

            colorAttachment.storeAction = MTLStoreActionStore;

            id<MTLRenderCommandEncoder>  mtlEncoder =
                [ctx->mtlCommandBuffer renderCommandEncoderWithDescriptor:ctx->mtlRenderPassDesc];
            MTLViewport vp = {0, 0, ctx->mtlFrameBuffer.width, ctx->mtlFrameBuffer.height, 0, 1};
            //fprintf(stderr, "%f %f \n", self.drawableSize.width, self.drawableSize.height);
            [mtlEncoder setViewport:vp];
            [mtlEncoder setRenderPipelineState:ctx->mtlBlitPipelineState];
            [mtlEncoder setVertexBuffer:ctx->mtlUniformBuffer
                              offset:0 atIndex:FrameUniformBuffer];

            [mtlEncoder setFragmentTexture: ctx->mtlFrameBuffer atIndex: 0];
            [mtlEncoder setVertexBuffer:ctx->mtlVertexBuffer offset:0 atIndex:MeshVertexBuffer];
            [mtlEncoder drawPrimitives:MTLPrimitiveTypeTriangle vertexStart:0 vertexCount:N * 3];
            [mtlEncoder endEncoding];

            [ctx->mtlRenderPassDesc release];
            ctx->mtlRenderPassDesc = nil;
        }

        if (!ctx->mtlEmptyCommandBuffer) {
            [ctx->mtlCommandBuffer presentDrawable:mtlDrawable];
            [ctx->mtlCommandBuffer commit];
        }


        [ctx->mtlCommandBuffer release];
        ctx->mtlCommandBuffer = nil;
        ctx->mtlEmptyCommandBuffer = YES;
    }

    if (ctx->mtlDrawable) [ctx->mtlDrawable release];
    ctx->mtlDrawable = nil;
}

- (void) dealloc {
    self.javaLayer = nil;
    [super dealloc];
}

@end

/*
 * Class:     sun_java2d_metal_CGLLayer
 * Method:    nativeCreateLayer
 * Signature: ()J
 */
JNIEXPORT jlong JNICALL
Java_sun_java2d_metal_MTLLayer_nativeCreateLayer
(JNIEnv *env, jobject obj)
{
    __block MTLLayer *layer = nil;

JNF_COCOA_ENTER(env);

    JNFWeakJObjectWrapper *javaLayer = [JNFWeakJObjectWrapper wrapperWithJObject:obj withEnv:env];

    [ThreadUtilities performOnMainThreadWaiting:YES block:^(){
            AWT_ASSERT_APPKIT_THREAD;
        
            layer = [[MTLLayer alloc] initWithJavaLayer: javaLayer];
    }];
    
JNF_COCOA_EXIT(env);

    return ptr_to_jlong(layer);
}

// Must be called under the RQ lock.
JNIEXPORT void JNICALL
Java_sun_java2d_metal_MTLLayer_validate
(JNIEnv *env, jclass cls, jlong layerPtr, jobject surfaceData)
{
    MTLLayer *layer = OBJC(layerPtr);
fprintf(stderr, "Java_sun_java2d_metal_MTLLayer_validate\n");
    if (surfaceData != NULL) {
        BMTLSDOps *bmtlsdo = (BMTLSDOps*) SurfaceData_GetOps(env, surfaceData);
        layer.bufferWidth = bmtlsdo->width;
        layer.bufferHeight = bmtlsdo->height;
        layer.ctx = (MTLCtxInfo*)(((MTLSDOps *)bmtlsdo->privOps)->configInfo->context->ctxInfo);
        layer.device = layer.ctx->mtlDevice;
    } else {
        layer.ctx = NULL;
    }
}

// Must be called on the AppKit thread and under the RQ lock.
JNIEXPORT void JNICALL
Java_sun_java2d_metal_MTLLayer_blitTexture
(JNIEnv *env, jclass cls, jlong layerPtr)
{
    fprintf(stderr, "Java_sun_java2d_metal_MTLLayer_blitTexture\n");
    MTLLayer *layer = jlong_to_ptr(layerPtr);

    [layer blitTexture];
}

JNIEXPORT void JNICALL
Java_sun_java2d_metal_MTLLayer_nativeSetScale
(JNIEnv *env, jclass cls, jlong layerPtr, jdouble scale)
{
    JNF_COCOA_ENTER(env);
    MTLLayer *layer = jlong_to_ptr(layerPtr);
    // We always call all setXX methods asynchronously, exception is only in 
    // this method where we need to change native texture size and layer's scale
    // in one call on appkit, otherwise we'll get window's contents blinking, 
    // during screen-2-screen moving.
    [ThreadUtilities performOnMainThreadWaiting:[NSThread isMainThread] block:^(){
        layer.contentsScale = scale;
    }];
    JNF_COCOA_EXIT(env);
}

File: src/java.desktop/macosx/classes/sun/java2d/metal/MTLLayer.java
/*
 * Copyright (c) 2019, 2021, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package sun.java2d.metal;

import sun.awt.AWTAccessor;
import sun.java2d.NullSurfaceData;
import sun.java2d.SurfaceData;
import sun.lwawt.LWWindowPeer;
import sun.lwawt.macosx.CFLayer;
import sun.util.logging.PlatformLogger;

import java.awt.Component;
import java.awt.GraphicsConfiguration;
import java.awt.Insets;
import java.awt.Window;


public final class MTLLayer extends CFLayer {
    private static final PlatformLogger logger = PlatformLogger.getLogger(MTLLayer.class.getName());

    private native long nativeCreateLayer(boolean perfCountersEnabled);
    private static native void nativeSetScale(long layerPtr, double scale);

    // Pass the insets to native code to make adjustments in blitTexture
    private static native void nativeSetInsets(long layerPtr, int top, int left);
    private static native void validate(long layerPtr, MTLSurfaceData mtlsd);
    private static native void blitTexture(long layerPtr);
    private static native void nativeSetOpaque(long layerPtr, boolean opaque);

    private int scale = 1;

    public MTLLayer(LWWindowPeer peer) {
        super(0, true);

        Window target = (peer != null) ? peer.getTarget() : null;
        boolean perfCountersEnabled = (target != null) && AWTAccessor.getWindowAccessor().countersEnabled(target);

        setPtr(nativeCreateLayer(perfCountersEnabled));
        this.peer = peer;

        MTLGraphicsConfig gc = (MTLGraphicsConfig)getGraphicsConfiguration();
        if (logger.isLoggable(PlatformLogger.Level.FINE)) {
            logger.fine("device = " + (gc != null ? gc.getDevice() : "null"));
        }
        if (gc != null) {
            setScale(gc.getDevice().getScaleFactor());
        }
    }

    @Override
    public SurfaceData replaceSurfaceData(int scale) {
        if (getBounds().isEmpty()) {
            surfaceData = NullSurfaceData.theInstance;
            return surfaceData;
        }

        // the layer redirects all painting to the buffer's graphics
        // and blits the buffer to the layer surface (in display callback)
        MTLGraphicsConfig gc = (MTLGraphicsConfig)getGraphicsConfiguration();
        surfaceData = gc.createSurfaceData(this);
        if (scale <= 0) {
            scale = gc.getDevice().getScaleFactor();
        }
        setScale(scale);
        if (peer != null) {
            Insets insets = peer.getInsets();
            execute(ptr -> nativeSetInsets(ptr, insets.top, insets.left));
        }
        // the layer holds a reference to the buffer, which in
        // turn has a reference back to this layer
        if (surfaceData instanceof MTLSurfaceData) {
            validate((MTLSurfaceData)surfaceData);
        }

        return surfaceData;
    }

    public void validate(final MTLSurfaceData mtlsd) {
        MTLRenderQueue rq = MTLRenderQueue.getInstance();
        rq.lock();
        try {
            execute(ptr -> validate(ptr, mtlsd));
        } finally {
            rq.unlock();
        }
    }

    @Override
    public void dispose() {
        // break the connection between the layer and the buffer
        validate(null);
        SurfaceData oldData = surfaceData;
        surfaceData = NullSurfaceData.theInstance;
        if (oldData != null) {
            oldData.flush();
        }
        super.dispose();
    }

    private void setScale(final int _scale) {
        if (scale != _scale) {
            if (logger.isLoggable(PlatformLogger.Level.FINE)) {
                MTLGraphicsConfig gc = (MTLGraphicsConfig)getGraphicsConfiguration();
                logger.fine("current scale = " + scale + ", new scale = " + _scale + " (device = " + (gc != null ? gc.getDevice() : "null") + ")");
            }
            scale = _scale;
            execute(ptr -> nativeSetScale(ptr, scale));
        }
    }

    @Override
    public void setOpaque(boolean opaque) {
        execute(ptr -> nativeSetOpaque(ptr, opaque));
    }
// ----------------------------------------------------------------------
    // NATIVE CALLBACKS
    // ----------------------------------------------------------------------

    private void drawInMTLContext() {
        // tell the flusher thread not to update the intermediate buffer
        // until we are done blitting from it
        MTLRenderQueue rq = MTLRenderQueue.getInstance();
        rq.lock();
        try {
            execute(ptr -> blitTexture(ptr));
        } finally {
            rq.unlock();
        }
    }

    private void countNewFrame() {
        // Called from the native code when this layer has been presented on screen
        Component target = peer.getTarget();
        if (target instanceof Window window) {
            AWTAccessor.getWindowAccessor().bumpCounter(window, "java2d.native.frames");
        }
    }

    private void countDroppedFrame() {
        // Called from the native code when an attempt was made to present this layer
        // on screen, but that attempt was not successful. This can happen, for example,
        // when those attempts are too frequent.
        Component target = peer.getTarget();
        if (target instanceof Window window) {
            AWTAccessor.getWindowAccessor().bumpCounter(window, "java2d.native.framesDropped");
        }
    }
}
